### PR TITLE
Node and Path walk performance improvement

### LINKED ISSUE #7698: [18NewEngland] Game becomes unresponsive
**What's your Game ID?**
68247

https://18xx.games/game/68247

**How to Reproduce the Bug?**

What steps cause the unexpected behavior?
Trying to lay track on my turn, The page becomes unresponsive for a while and then doesnt accept my the track I layed.
I tried several different locations for my track but it didn't make a difference
Operating Round 9.2. The bank has broken

Please include screenshots (you can drag and drop or upload them into this issue)
image

**Which Browser / Platform?**
On Chrome 100.0.4896.127

### PR DESCRIPTION
**Pull request by roseundy**
Fixes #7698 

For the above New England game, this results in a modest speedup of loading the game (34 sec down to 19 sec)

* It now automatically detects cases where nodes and paths can be pruned instead of requiring a hint that Lawson tiles are used. It can find places in non-Lawson games where this pruning can happen.

### CODE REVIEW

**Comment by roseundy** [comment]
Short slide deck with some explanations:
[Walk bitvectors.pdf](https://github.com/tobymao/18xx/files/8711042/Walk.bitvectors.pdf)

**Comment by dfannius** [comment]
Could there be a new bitvector class that both Node and Path use? The concept is pretty general and the new code added to node.rb and path.rb has a lot of duplication.

**Comment by roseundy** [comment]
Yes, Toby and I are iterating on this. I wanted to get this in front of him, probably before it's fully baked.

File: lib/engine/bitvector.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      @vector[target.bit_group] = (@vector[target.bit_group] | (1 << target.bit_index)) ^ (1 << target.bit_index)
+    end
+
+    private
+
+    def check_and_allocate(target)
+      if target.walk_graph != @graph || target.walk_id != @id
+        target.walk_graph = @graph
+        target.walk_id = @id
+        target.bit_group = @next_index.div(BITS_PER_GROUP)
```
Comment (tobymao) [inline lib/engine/bitvector.rb:36]: why do you use div and not / ?

Comment (roseundy) [reply lib/engine/bitvector.rb:36]: Just more explicit that it is integer division, I guess. Is it a problem?

Comment (roseundy) [reply lib/engine/bitvector.rb:36]: I'll change it

**Comment by roseundy** [comment]
Updated slide deck:

[Walk bitvectors.pdf](https://github.com/tobymao/18xx/files/8722886/Walk.bitvectors.pdf)

File: lib/engine/bit_vector.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    end
+
+    def [](target)
+      check_and_allocate(target)
+      ((@vector[target.bit_group] >> target.bit_index) & 1) == 1
+    end
+
+    def []=(target, val)
+      check_and_allocate(target)
+      bit = val ? 1 : 0
```
Comment (tobymao) [inline lib/engine/bit_vector.rb:21]: is val ever not true?

File: lib/engine/bit_vector.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      @vector[target.bit_group] |= (bit << target.bit_index)
+    end
+
+    def delete(target)
+      check_and_allocate(target)
+      @vector[target.bit_group] = (@vector[target.bit_group] | (1 << target.bit_index)) ^ (1 << target.bit_index)
```
Comment (tobymao) [inline lib/engine/bit_vector.rb:27]: couldn't this just be 

self[target] = false

File: lib/engine/bit_vector.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    private
+
+    def check_and_allocate(target)
+      if target.walk_graph != @graph || target.walk_id != @id
+        target.walk_graph = @graph
+        target.walk_id = @id
+        target.bit_group = @next_index / BITS_PER_GROUP
+        target.bit_index = @next_index % BITS_PER_GROUP
+        @next_index += 1
```
Comment (tobymao) [inline lib/engine/bit_vector.rb:38]: whoa this is strange and unexpected, why are you incrementing next_index based on if it's a new bit vector, shouldn't you increment next_index when you overflow it?

File: lib/engine/tile.rb [hunk 5]
```diff
@@ -288,6 +289,21 @@ def exits
       @_exits ||= @edges.map { |e| rotate(e.num, @rotation) }.uniq
     end
 
+    def converging_exit?(num)
+      exit_count[num] > 1
+    end
+
+    def exit_count
+      @_exit_count ||= compute_exit_count
```
Comment (tobymao) [inline lib/engine/tile.rb:297]: just make this inline

@_exit_count ||= begin
    counts = Hash.new(0) 
   @edges.each { |edge| counts[...) += 1 }
   counts

File: lib/engine/tile.rb [hunk 6]
```diff
@@ -288,6 +289,21 @@ def exits
       @_exits ||= @edges.map { |e| rotate(e.num, @rotation) }.uniq
     end
 
+    def converging_exit?(num)
+      exit_count[num] > 1
```
Comment (tobymao) [inline lib/engine/tile.rb:293]: (exit_count[num] || 0) > 1

**Pull request merged**